### DISCUSSION TITLE
Allow additional parameters in Route() to be added to scope

### DISCUSSION THREAD

**Discussion by dhait** [encode/starlette#2198] [Ideas]
For some use-cases (ex.  Model-View-Controller), we might need more flexibility in the routing.  For example, we might have several "GET" routes all going to the same Endpoint, but with different methods (depending on the particular path or path parameters).

An easy way to do this would be to allow the addition of user-chosen keywords to the Route class:

```
routes = [
      Route('/abc', MyEndpoint, methods=['GET'], action='list'),
      Route('/abc/{id}', MyEndpoint, methods=['GET'], action='find')
]
```

These keywords could then be passed to the Endpoint by simply adding them to the scope, i.e. in the handle() function:

    `scope['action'] = 'list'
     await response(scope, receive, send)`

Now the endpoint can pull this from the scope and modify how it behaves.

If you prefer, instead of keywords, an additional `context` dictionary could be passed and added into the scope that way:

`Route('/abc', MyEndpoint, methods=['GET'], context={'action' : 'list', 'authenticated' : True} )`

What is your opinion?
An easy PR, I would think.

**Comment by alex-oleshkevich** [comment]
`action` won't take any effect when you use functional views.
I think the thing you describe is a different route type:

```python
class Controller:
    def __init__(self, endpoint: ASGIApp):
     	self.endpoint = endpoint

    async def __call__(self, scope, receive, send):
		method = scope['method']
		# do here any internal routing 
		# eg. map POST to ` self.endpoint.create` or map GET to `self.endpoint.find`

routes = [
	Controller('/users', MyEndpoint),
]
```

**Reply by dhait** [reply]
Your example shows the instance method is tied to the HTTP method.  I am instead proposing that we decouple the HTTP method from the endpoint method itself.  As I stated, this is extremely helpful in the event you want to use one endpoint for **two different** kind of GET routes.

How else would you do this? I have a controller in charge of "Notes".  How would I route "GET /notes" and "GET /notes/new" to two different controller methods?  

`if method == 'GET' and 'new' in path:  # etc for each possible combination`

This would require significant extra logic in the code you show above.  I am simply allowing a shortcut here.

As you say, this is mainly targeted for HTTPEndpoints.  But from a practical perspective, functional views already have this capability: you can name the function anything you want.  I am suggesting we allow the same flexibility for class endpoints.  My proposed approach is flexible enough to allow it to be used only when needed.
